### PR TITLE
Tray QoL

### PR DESCRIPTION
**Pull request by Geevies**
* You can now remove individual items from a tray by having it in one hand, and clicking it with the other empty hand.
* The sprites on trays are now randomly spread out instead of just going into the centre.

### CODE REVIEW

**Comment by Geevies** [comment]
!review

File: code/game/objects/items/weapons/trays.dm [hunk 1]
```diff
@@ -178,162 +168,153 @@
...
+		if(addedSomething == 1)
+			user.visible_message(SPAN_NOTICE("\The [user] loads an item onto their service tray."))
+		else if (addedSomething)
+			user.visible_message(SPAN_NOTICE("\The [user] loads [addedSomething] items onto their service tray."))
 		else
-			to_chat(user, "The tray is full or there's nothing valid here")
-			return 1
-		return 0//This prevents the alt-click from doing any farther actions
-	return 1
+			to_chat(user, SPAN_WARNING("The tray is full or there's nothing valid here."))
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]: "there's nothing valid here" is a really awkward message.

File: code/game/objects/items/weapons/trays.dm [hunk 2]
```diff
@@ -178,162 +168,153 @@
...
+			if(attempt_load_item(I, usr, 0))
 				addedSomething++
-		if ( addedSomething == 1)
-			usr.visible_message("<span class='notice'>[user] loads an item onto their service tray.</span>")
-		else if ( addedSomething )
-			usr.visible_message("<span class='notice'>[user] loads [addedSomething] items onto their service tray.</span>")
+		if(addedSomething == 1)
+			user.visible_message(SPAN_NOTICE("\The [user] loads an item onto their service tray."))
+		else if (addedSomething)
+			user.visible_message(SPAN_NOTICE("\The [user] loads [addedSomething] items onto their service tray."))
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]: These can probably be condensed and made less awkward.
```suggestion
		if(addedSomething)
			user.visible_message(SPAN_NOTICE("\The [user] loads [addedsomething > 1 ? "some things" : I.name] onto their service tray."))
```

File: code/game/objects/items/weapons/trays.dm [hunk 3]
```diff
@@ -178,162 +168,153 @@
...
 		else
-			to_chat(user, "The tray is full or there's nothing valid here")
-			return 1
-		return 0//This prevents the alt-click from doing any farther actions
-	return 1
+			to_chat(user, SPAN_WARNING("The tray is full or there's nothing valid here."))
+			return TRUE
+		return FALSE //This prevents the alt-click from doing any farther actions
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]: god i know it's nitpicky and not something you added but this is driving me crazy
```suggestion
		return FALSE //This prevents the alt-click from doing any further actions
```

File: code/game/objects/items/weapons/trays.dm [hunk 4]
```diff
@@ -178,162 +168,153 @@
...
 
 /obj/item/tray/verb/unload()
 	set name = "Unload Tray"
 	set category = "Object"
 	set src in view(1)
 
-	if (!istype(loc,/turf))//check that we're not being held by a mob
-		to_chat(usr, "Place the tray down first!")
+	if(!istype(loc, /turf))//check that we're not being held by a mob
+		to_chat(usr, SPAN_WARNING("Place the tray down first!"))
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]: Should consider replacing instances of "the tray" with "\the [src]"
```suggestion
		to_chat(usr, SPAN_WARNING("Place \the [src] down first!"))
```

File: code/game/objects/items/weapons/trays.dm [hunk 5]
```diff
@@ -178,162 +168,153 @@
...
-
+		var/turf/dropspot = get_turf(src)
 		for(var/obj/item/I in carrying)
 			I.forceMove(dropspot)
 			carrying -= I
-
 		cut_overlays()
 		current_weight = 0
-		usr.visible_message("[usr] unloads the tray.", "You unload the tray.")
+		usr.visible_message(SPAN_NOTICE("[usr] unloads the tray."), SPAN_NOTICE("You unload the tray."))
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]: As above

File: code/game/objects/items/weapons/trays.dm [hunk 6]
```diff
@@ -178,162 +168,153 @@
...
 
 /obj/item/tray/proc/unload_at_loc(var/turf/dropspot = null, var/mob/user)
-	if (!istype(loc,/turf) && !dropspot)//check that we're not being held by a mob
-		to_chat(user, "Place the tray down first!")
+	if(!istype(loc, /turf) && !dropspot)//check that we're not being held by a mob
+		to_chat(user, SPAN_WARNING("Place the tray down first!"))
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]: As above

File: code/game/objects/items/weapons/trays.dm [hunk 7]
```diff
@@ -178,162 +168,153 @@
...
+			dropspot = get_turf(src)
 
 		for(var/obj/item/I in carrying)
 			I.forceMove(dropspot)
 			carrying -= I
 
 		cut_overlays()
 		current_weight = 0
-		user.visible_message("[user] unloads the tray.", "You unload the tray.")
+		user.visible_message(SPAN_NOTICE("\The [user] unloads the tray."), SPAN_NOTICE("You unload the tray."))
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]: As above

File: code/game/objects/items/weapons/trays.dm [hunk 8]
```diff
@@ -178,162 +168,153 @@
...
 			I.forceMove(dropspot)
 			carrying.Remove(I)
 			spawn()
 				for(var/i = 1, i <= rand(1,2), i++)
 					if(I)
-						step(I, pick(NORTH,SOUTH,EAST,WEST))
-						sleep(rand(2,4))
-		if (user)
-			user.visible_message("<span class='notice'>[user] spills their tray all over the floor.</span>")
+						step(I, pick(NORTH, SOUTH, EAST, WEST))
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]:
```suggestion
						step(I, pick(cardinal))
```

File: code/game/objects/items/weapons/trays.dm [hunk 9]
```diff
@@ -178,162 +168,153 @@
...
-	//rand(0, (max_offset_y*2)-3)-(max_offset_y)-3
 
 /obj/item/tray/verb/unload()
 	set name = "Unload Tray"
 	set category = "Object"
 	set src in view(1)
 
-	if (!istype(loc,/turf))//check that we're not being held by a mob
-		to_chat(usr, "Place the tray down first!")
+	if(!istype(loc, /turf))//check that we're not being held by a mob
```
Comment (Doxxmedearly) [inline code/game/objects/items/weapons/trays.dm]:
```suggestion
	if(!isturf(loc)) //check that we're not being held by a mob
```

File: code/game/objects/items/weapons/trays.dm [hunk 10]
```diff
@@ -178,162 +168,153 @@
...
+					if(messages)
+						to_chat(user, SPAN_NOTICE("You place \the [I] on \the [src]."))
+					return TRUE
 				else
-					if (messages)
-						to_chat(user, "The tray can't take that much weight")
-		if (!match && messages)to_chat(user, "That item isn't suitable for a tray")
-	return 0
+					if(messages)
+						to_chat(user, SPAN_WARNING("\The [src] can't take hold that much weight."))
```
Comment (skull132) [inline code/game/objects/items/weapons/trays.dm]: Why doesn't this break? It should.

File: code/game/objects/items/weapons/trays.dm (hunk 3 above)
Comment (Geevies) [reply code/game/objects/items/weapons/trays.dm]: father......

File: code/game/objects/items/weapons/trays.dm [hunk 11]
```diff
@@ -34,88 +35,71 @@
...
-			for(var/mob/O in viewers(M, null))
-				O.show_message(text("<span class='danger'>[] slams [] with the tray!</span>", user, M), 1)
-			return
-		else
-			playsound(M, 'sound/items/trayhit2.ogg', 50, 1)  //we applied the damage, we played the sound, we showed the appropriate messages. Time to return and stop the proc
-			for(var/mob/O in viewers(M, null))
-				O.show_message(text("<span class='danger'>[] slams [] with the tray!</span>", user, M), 1)
-			return
+		playsound(M, pick('sound/items/trayhit1.ogg', 'sound/items/trayhit2.ogg'), 50, TRUE)
+		visible_message("\The [user] slams \the [M] with \the [src]!")
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/trays.dm]: Why does this directly give organ damage? 🤔 
Also maybe make it slam people in the organ they actually did.

File: code/game/objects/items/weapons/trays.dm [hunk 12]
```diff
@@ -34,88 +35,71 @@
...
-				O.show_message(text("<span class='danger'>[] slams [] with the tray!</span>", user, M), 1)
-		else
-			playsound(M, 'sound/items/trayhit2.ogg', 50, 1)  //sound playin'
-			for(var/mob/O in viewers(M, null))
-				O.show_message(text("<span class='danger'>[] slams [] with the tray!</span>", user, M), 1)
+		playsound(M, pick('sound/items/trayhit1.ogg', 'sound/items/trayhit2.ogg'), 50, 1)
+		visible_message(SPAN_DANGER("\The [user] slams \the [M] with the tray!"))
 		if(prob(10))
-			if(!istype(M,/mob/living/silicon))M.Stun(rand(1,3))
+			if(!istype(M,/mob/living/silicon))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/trays.dm]:
```suggestion
			if(!issilicon(M))
```

File: code/game/objects/items/weapons/trays.dm [hunk 13]
```diff
@@ -149,25 +126,38 @@
 				return
 			return
 
-/obj/item/tray/var/cooldown = 0	//shield bash cooldown. based on world.time
-
-/obj/item/tray/attackby(obj/item/W as obj, mob/user as mob)
-	if (istype(user,/mob/living/silicon/robot))//safety to stop robots losing their items
+/obj/item/tray/attackby(obj/item/W, mob/user)
+	if(istype(user,/mob/living/silicon/robot)) //safety to stop robots losing their items
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/trays.dm]:
```suggestion
	if(isrobot(user)) //safety to stop robots losing their items
```

File: code/game/objects/items/weapons/trays.dm (hunk 11 above)
Comment (Geevies) [reply code/game/objects/items/weapons/trays.dm]: is that in scope

Comment (NonQueueingMatt) [reply code/game/objects/items/weapons/trays.dm]: Depends on you.

File: code/game/objects/items/weapons/trays.dm [hunk 14]
```diff
@@ -178,162 +168,152 @@
...
 
 /obj/item/tray/throw_at(atom/target, throw_range, throw_speed, mob/user)
-	safedrop = 1//we dont want the tray to spill when thrown, it will spill on impact instead
+	safedrop = TRUE //we dont want the tray to spill when thrown, it will spill on impact instead
 	..()
 
 /obj/item/tray/dropped(mob/user)
 	spawn(1)//A hack to avoid race conditions. Dropped procs too quickly
-		if (istype(src.loc, /mob))
+		if(istype(src.loc, /mob))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/trays.dm]:
```suggestion
		if(ismob(loc))
```

**Pull request merged**